tensorflow/compiler/tf2tensorrt/convert/ops/variable_ops.cc

```diff
@@ -0,0 +1,209 @@
+#include "tensorflow/compiler/tf2tensorrt/common/utils.h"
+#include "tensorflow/compiler/tf2tensorrt/convert/convert_nodes.h"
+#include "tensorflow/compiler/tf2tensorrt/convert/op_converter.h"
+#include "tensorflow/compiler/tf2tensorrt/convert/op_converter_registry.h"
+#include "tensorflow/compiler/tf2tensorrt/convert/utils.h"
+#include "tensorflow/core/common_runtime/process_function_library_runtime.h"
+#include "tensorflow/core/lib/core/errors.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/platform/stream_executor.h"
+#include "third_party/tensorrt/NvInfer.h"
+#include "third_party/tensorrt/NvInferRuntimeCommon.h"
+
+namespace tensorflow {
+namespace tensorrt {
+namespace convert {
+
+class ConvertVariableV2 : public OpConverterBase<ConvertVariableV2> {
+ public:
+  ConvertVariableV2(OpConverterParams* params)
+      : OpConverterBase<ConvertVariableV2>(params) {}
+
+  struct VariableV2Attributes {
+    TensorShapeProto shape_proto;
+    TensorShape shape;
+    string name;
+    DataType dtype;
+    string shared_name;
+    string container;
+  };
+
+  static constexpr std::array<InputArgSpec, 0> InputSpec() {
+    return {};
+  }
+
+  static constexpr std::array<DataType, 2> AllowedDataTypes() {
+    return {DataType::DT_FLOAT, DataType::DT_HALF};
+  }
+
+  static constexpr const char* NodeDefDataTypeAttributeName() {
+    return "dtype";
+  }
+
+  template <typename T>
+  Status ValidateAux() {
+    const auto& node_def = params_->node_def;
+
+    // Verify and consume node attributes.
+    StatusOr<TensorShapeProto> shape_proto = GetAttrValue<TensorShapeProto>("shape");
+    StatusOr<string> shared_name = GetAttrValue<string>("shared_name");
+    StatusOr<string> container = GetAttrValue<string>("container");
+    TRT_ENSURE_OK(shape_proto);
+    TRT_ENSURE_OK(shared_name);
+    TRT_ENSURE_OK(container);
+
+    attrs_.shape_proto = *shape_proto;
+    attrs_.shape = TensorShape(*shape_proto);
+    attrs_.name = node_def.name();
+    attrs_.shared_name = *shared_name;
+    attrs_.container = *container;
+
+    Tensor tensor(attrs_.dtype, attrs_.shape);
+    auto tensor_flat = tensor.flat<T>();
+    for (int64_t i = 0; i < tensor_flat.size(); i++) {
+      tensor_flat(i) = T(0.0f);
+    }
+
+    TRT_ShapedWeights weights;
+    TF_RETURN_IF_ERROR(
+        TfTensorToTrtWeights(tensor, params_->weight_store, &weights));
+
+    // Only push outputs during validation and when outputs are expected.
+    if (params_->validation_only && params_->outputs != nullptr) {
+      AddOutput(TRT_TensorOrWeights(weights));
+    }
+    return Status::OK();
+  }
+
+  Status Validate() {
+    StatusOr<DataType> dtype = GetAttrValue<DataType>("dtype");
+    TRT_ENSURE_OK(dtype);
+    attrs_.dtype = *dtype;
+
+    switch (attrs_.dtype) {
+      case DT_FLOAT:
+        return ValidateAux<float>();
+      case DT_HALF:
+        return ValidateAux<Eigen::half>();
+    }
+  }
+
+  template <typename T>
+  Status ConvertAux() {
```

Comment:
So the way we get the VariableV2 value from the OpKernelContext is to execute a kernel with just the VariableV2 on GPU then copy the value to host? Can we get the value by executing the same kernel on host, without the copy?

Reply:
I think the reason why the variable is on the GPU is that the context we're using here is that of the TRT engine op and the device is the GPU. Maybe I'm wrong, but my understanding is that the tensor in the resource manager is on the GPU and the copy is unavoidable.